feat/data_passing

alhiwatan opened the pull request:
### Add support for data passing

https://app.asana.com/0/337926977135969/1202164609786776/f

- based on the Unity plugin for overall design and APIs
- used camelcase for the field names (APIs), which is different from Unity,  the APIs are:

```
AppLovinMAX.TargetingData.yearOfBirth
AppLovinMAX.TargetingData.gender
AppLovinMAX.TargetingData.maximumAdContentRating
AppLovinMAX.TargetingData.email
AppLovinMAX.TargetingData.phoneNumber
AppLovinMAX.TargetingData.keywords
AppLovinMAX.TargetingData.interests
AppLovinMAX.TargetingData.clearAll();
```

**Done following tests**

- Set and unset data on the fly during the execution on both Android and iOS, verified by Charles, the output is

```
"targeting_data": {
	"gender": "M",
	"keywords": "fruit:apple,fruit:banana,fruit:orange",
	"interests": "drawing,painting,singing,dancing",
	"year_of_birth": "2000",
	"email": "a6ad00ac113a19d953efb91820d8788e2263b28a",
	"phone_number": "411d5ce79499b451aa395897ef1a63a705e18f1d",
	"maximum_ad_content_rating": "3"
},
```

- Clear the data one by one as documented.

```
AppLovinMAX.TargetingData.yearOfBirth = 0;
AppLovinMAX.TargetingData.gender = AppLovinMAX.UserGender.Unknown;
AppLovinMAX.TargetingData.maximumAdContentRating = AppLovinMAX.AdContentRating.None;
AppLovinMAX.TargetingData.email = null;
AppLovinMAX.TargetingData.phoneNumber = null;
AppLovinMAX.TargetingData.keywords = null;
AppLovinMAX.TargetingData.interests = null;
```
- Error messages before SDK initialization

`2022-05-11 11:58:45.868488+0900 AppLovinMAX Example[71485:4462281] [AppLovinSdk] [AppLovinMAX] ERROR: Failed to execute setTargetingDataKeywords() - please ensure the AppLovin MAX React Native module has been initialized by calling 'AppLovinMAX.initialize(...);'!`

thomasmso commented on ios/AppLovinMAX.m:
```diff
@@ -206,6 +206,128 @@ - (instancetype)init
+    self.sdk.targetingData.gender = alGender;
+}
+
+RCT_EXPORT_METHOD(setTargetingDataMaximumAdContentRating:(nonnull NSNumber *)maximumAdContentRating)
```
We usually assume `nonnull` in our codebase, and only annotate if something is `nullable`.

alhiwatan replied:
Without nonnull, RN generates an error like:

`Argument 0 (NSNumber) of AppLovinMAX.setTargetingDataMaximumAdContentRating has unspecified nullability but React requires that all NSNumber arguments are explicitly marked as nonnull to ensure compatibility with Android`

We can use ‘double’ for passing numbers in RN but double is not straightforward for comparison.  So I think that nonnull is ok for this case.

thomasmso replied:
Ah ok, I remember that being an issue in RN.

thomasmso commented on ios/AppLovinMAX.m:
```diff
@@ -206,6 +206,128 @@ - (instancetype)init
+    
+    ALAdContentRating rating = ALAdContentRatingNone;
+
+    int intVal = [ maximumAdContentRating intValue ];
```
nit but in our codebase we usually use property notation
```suggestion
    int intVal = maximumAdContentRating.intValue;
```

thomasmso commented on ios/AppLovinMAX.m:
```diff
@@ -206,6 +206,128 @@ - (instancetype)init
+    self.sdk.targetingData.maximumAdContentRating = rating;
+}
+
+RCT_EXPORT_METHOD(setTargetingDataEmail:(NSString *)email)
```
We might need to let the pub be able to clear this (and other) values. 

Our iOS SDK API:

```
/**
 * The email of the user.
 * Set this property to @c nil to clear this value.
 */
@property (nonatomic, copy, nullable) NSString *email;
```

alhiwatan replied:
Ok, I'll add 'nullable' for those functions explicitly.

thomasmso commented on ios/AppLovinMAX.m:
```diff
@@ -923,6 +1045,11 @@ - (void)logInvalidAdFormat:(MAAdFormat *)adFormat
     [self log: @"invalid ad format: %@, from %@", adFormat, [NSThread callStackSymbols]];
 }
 
+- (void)logUninitializedAccessError:(NSString *)callingMethod
```
Should we port this to other API calls as well?

alhiwatan replied:
Ok, I'll update the Java side as well.

alhiwatan commented:
Updated the following items

- added nullable explicitly for the iOS functions
- used logUninitializedAccessError() for the other APIs
- fixed invalid property access on clearAll
      [self.sdk.targetingData.clearAll]; => [self.sdk.targetingData clearAll];

thomasmso commented on ios/AppLovinMAX.m:
```diff
@@ -923,6 +1045,11 @@ - (void)logInvalidAdFormat:(MAAdFormat *)adFormat
 
+- (void)logUninitializedAccessError:(NSString *)callingMethod
+{
+    [self log: @"ERROR: Failed to execute %@() - please ensure the AppLovin MAX React Native module has been initialized by calling 'AppLovinMAX.initialize(...);'!", callingMethod ];
```
```suggestion
    [self log: @"ERROR: Failed to execute %@() - please ensure the AppLovin MAX React Native module has been initialized by calling 'AppLovinMAX.initialize(...);'!", callingMethod];
```

thomasmso reviewed (approved):
Now we just have to create a subtask for David to create docs, then we can push

alhiwatan commented:
Let me ask for one more review.   I forgot to add setLocationCollectionEnabled() for Location Passing.

alhiwatan commented:
Added setLocationCollectionEnabled(), verified this to be added on true and removed on false.

```
	"location_info": {
		"loc_auth": 0,
		"loc_services_enabled": true
	},
```
